### PR TITLE
Implement deconstruction for LocalDate

### PR DESCRIPTION
**Pull request by carlosschults**
As suggested in #585.

### CODE REVIEW

File: src/NodaTime/LocalDate.cs [hunk 1]
```diff
@@ -746,6 +746,20 @@ public LocalDate Previous(IsoDayOfWeek targetDayOfWeek)
         public LocalDate With([NotNull] Func<LocalDate, LocalDate> adjuster) =>
             Preconditions.CheckNotNull(adjuster, nameof(adjuster)).Invoke(this);
 
+        /// <summary>
```
Comment (jskeet) [inline src/NodaTime/LocalDate.cs]: Possibly add another overload accepting a CalendarSystem (as a final parameter)?

File: src/NodaTime.Test/LocalDateTest.cs [hunk 2]
```diff
@@ -71,5 +71,22 @@ public void MinIsoValue()
...
+        {
+            var value = new LocalDate(2017, 11, 7);
+            var expectedYear = 2017;
+            var expectedMonth = 11;
+            var expectedDay = 7;
+            value.Deconstruct(out int actualYear, out int actualMonth, out int actualDay);
+
+            Assert.Multiple(() =>
+            {
+                Assert.AreEqual(expectedYear, actualYear, "Year comparison failed.");
```
Comment (jskeet) [inline src/NodaTime.Test/LocalDateTest.cs]: No need for the messages here - if there are failures, we can tell what's failed by the stack trace, or probably just by the numbers.

File: src/NodaTime.Test/LocalDateTest.cs [hunk 3]
```diff
@@ -71,5 +71,22 @@ public void MinIsoValue()
...
         }
+
+        [Test]
+        public void Deconstruction()
+        {
+            var value = new LocalDate(2017, 11, 7);
+            var expectedYear = 2017;
+            var expectedMonth = 11;
+            var expectedDay = 7;
+            value.Deconstruct(out int actualYear, out int actualMonth, out int actualDay);
```
Comment (jskeet) [inline src/NodaTime.Test/LocalDateTest.cs]: I'd use the deconstruction feature here to make it more obvious why we're doing this:

```csharp
var (actualYear, actualMonth, actualDay) = value;
```

Comment (carlosschults) [reply src/NodaTime.Test/LocalDateTest.cs]: Yeah, I try doing that, but then I'd have to add the System.ValueTuple package, and I wasn't sure I could.

And in that case, to which project should I add it? Just to the test project?

File: src/NodaTime.Test/LocalDateTest.cs (hunk 2 above)
Comment (carlosschults) [reply src/NodaTime.Test/LocalDateTest.cs]: Got it.

File: src/NodaTime/LocalDate.cs (hunk 1 above)
Comment (carlosschults) [reply src/NodaTime/LocalDate.cs]: Sounds good.

File: src/NodaTime.Test/LocalDateTest.cs (hunk 3 above)
Comment (jskeet) [reply src/NodaTime.Test/LocalDateTest.cs]: Yup - it's fine to have it in tests, but I don't want it as a dependency of Noda Time itself. (I believe fairly soon we won't need it at all for deconstruction to variables - I can find the Roslyn issue if you want.)

**Comment by carlosschults** [comment]
@jskeet Thanks for the review.

Regarding which types should we add desconstruction to, maybe we should discuss in the issue?

**Comment by jskeet** [comment]
Yup, absolutely - in fact, when I said "PR" I really meant "issue" :)

File: src/NodaTime/LocalDate.cs [hunk 4]
```diff
@@ -760,6 +760,22 @@ public void Deconstruct(out int year, out int month, out int day)
...
+        /// Deconstructs the current instance into its components.
+        /// </summary>
+        /// <param name="year">The year component of the date.</param>
+        /// <param name="month">The month component of the date.</param>
+        /// <param name="day">The day component of the date.</param>
+        /// <param name="calendar">The <see cref="CalendarSystem"/> associated with the date.</param>
+        [Pure]
+        public void Deconstruct(out int year, out int month, out int day, [NotNull]out CalendarSystem calendar)
```
Comment (jskeet) [inline src/NodaTime/LocalDate.cs]: I think it's probably worth removing the `[NotNull]` attribute for the moment - aside from anything else, at the moment the documentation generator would add an `ArgumentNullException` XML doc. Will need to modify that - we haven't had much in the way of out parameters before...

(It's fine to use "amend" and force-push for that - no need for another commit.)

**Comment by carlosschults** [comment]
Well, it seems I've broken a test by removing the `[NotNull]` attribute :/

**Comment by jskeet** [comment]
Will have a look when I get the chance, but I'm travelling today/tomorrow. I may just fix the doc generator :)

**Comment by jskeet** [comment]
Okay, tell you what - let's get this in now *with* the NotNull attribute (which means tests should pass) and then I'll be able to fix up the tools in a way which is easy to verify. So if you could just put back the NotNull, that'd be grand - and please accept my sincere apologies for all the messing around :)

**Pull request merged**